File: examples/durf.pl
#!/usr/local/bin/perl
# -*-Perl-*-
eval 'exec /usr/local/bin/perl -S $0 ${1+"$@"}'
 if 0;
'di';
'ig00';
#From jimbo@rcx1.ssd.csd.harris.com Wed Jun 23 10:17:25 1993
#From: jimbo@rcx1.ssd.csd.harris.com (Jim Winters)
#Newsgroups: comp.lang.perl
#Subject: durf: disc usage report formatter
#Date: 19 Jun 93 17:53:23 GMT
#Organization: Harris Computer Systems Division

#If you are like me, you find the output of du(1) a little hard to
#interpret.  I mean, the fields are simple enough to understand, but
#it's just hard to get the big picture if you're du'ing a large tree.

#durf is a perl script that that formats the output of du differently.
#"Durf" is for "disk usage report formatter", (really just because I
#like the name).

#I'm something of a novice perl programmer, so my perl code is probably
#not the best, but I've been using this script for a few months now
#with no problems or even any desire to change it!  :-)

#I don't know how portable this will be to other systems, but I don't
#particularly expect any problems.  (I'm on a flavor of unix.)  Du on
#our system has a pretty simple format, and durf doesn't really care
#where the input comes from (if you use the '-' argument), as long as
#it's in that format.  I have noticed that durf does not work with some
#older versions of perl.

#The script has had the man page thingie (forgot the name, sorry)
#applied to it, so it is its own man page.  Also, durf has a -H option
#for help.

#Go to some (small) directory and type durf.  The output will look
#something like this (from my bin directory):


#Tot%  Blocks  Full-pathname
#    Tot% Sub% Blocks    Child

#100%    5019  .                     <--- name of top level directory
#     75%  75%   3804    m88k            <--- a child, "m88k"
#     15   15     779    .               <--- special child, "."
#      4    4     243    hcx
#      2    2     120    gcx
#      1    1      73    sun

# 75%    3804  ./m88k
#     75% 100%   3804    .

#(I've truncated the output for this article.)

#Basically durf is telling you how the space is distributed in the
#tree.  Here 75% of the total space is in the m88k subtree.  m88k also
#has 75% of the space in the subdirectory.  (Since this is the top
#level, all the percentages are the same.)  The other numbers are sizes
#in blocks.

#The special child "." is the amount of space in the directory
#exclusive of the children.

#After the list of top-level children, we see a listing for the m88k
#subtree.  Within this subtree, we see that 75% of the total space is
#in ".", which is consistent with the previous entry.  However, since
#m88k has no children, 100% of the space in THIS subtree is in ".".

#Naturally, this is all more interesting with bigger trees.

#durf requires "yagrip.pl" by jgreely@cis.ohio-state.edu, 89/11/1, so I
#have included it as well.  It needs to go in your perl library.

#I release this code for any purpose whatsoever.  Enjoy.

#
# $Header: /net/pauillac/caml/repository/cash/examples/durf.pl,v 1.1 2002/06/19 14:38:15 verlyck Exp $
#
# $Log: durf.pl,v $
# Revision 1.1  2002/06/19 14:38:15  verlyck
# Added the perl script from which was written durf.ml, with its motivation and
# documentation.
#

# durf - disk usage report formatter.
#

sub Usage {
    die &usage( @Usage ) , <<EndOfUsage;

-H      Write this message and die.
-t threshold[%]
        Only show directories at or above threshold blocks.  With "%",
        only show directories at or above threshold percent of the
        total.  Default is "1%"
du-args E.g., directory on which to operate.  Default is ".".
-       Read standard input for 'du' listing.

Like 'du', but formatted differently.
EndOfUsage
}

#
# Deal with options
#

$threshold = "1%";

# yagrip doit etre au meme endroit que durf
$_ = $ENV{'HOME'} . '/pl';
($_ = $0) =~ s|/[^/]*$|| unless -d $_; # le dir de ce programme-ci
unshift(@INC, $_);		# avec push, do merde quelquefois (??)

require "yagrip.pl";            # allows "--" to terminate options

$Option_string = "Ht:";

@Usage = ( &sname( $0 ), $Option_string, "threshold[%]", "[{du-args|-}]" );

&getopt( $Option_string ) || die &usage( @Usage );

if ( defined( $opt_H ) ) { &Usage(); }

if ( defined( $opt_t ) ) { $threshold = $opt_t; }

if ( $threshold =~ /%$/ ) {
    $threshold_is_percent = 1;
    chop( $threshold );
}

$threshold += 0;

#
# Prepare file handle
#
if ( $ARGV[0]  && $ARGV[0] eq '-' ) {
    $open_string = '-';
} else {
    $open_string = 'du ' . join( ' ', @ARGV ) . '|';
}

open( DU, $open_string ) || die "Can't open \"$open_string\".\n";

#
# Load input
#
while(<DU>) {
    chop;
    if ( /^\s*(\d+)\s+(\S.*)$/ ) {
        &load( $1, $2 );
    } else {
        print STDERR "Can't understand line:\n$_\n";
    }
}

#
# Establish list of roots
#
foreach (keys %size) {
    push( @root_list, $_ ) if ( ! $parent{$_} || ! $size{ $parent{$_} } );
}

#
# Calculate total size
#
grep( $total_size += $size{$_}, @root_list );

&print_headers();

#
# Output formatted tree for each root
#
while( $_ = pop( @root_list ) ) {
    &do_subtree( $_ );
}

exit( 0 );

sub do_subtree {
    push( @subtree_list, @_ );
    while( $_ = pop( @subtree_list ) ) {
        &do_output( $_ );
    }
}

sub sname {
    local( $name ) = @_;
    local( @name_array ) = split( '/', $name );
    pop( @name_array );
}

sub dname {
    local( $name ) = @_;
    local( @name_array ) = split( '/', $name );
    if ( scalar( @name_array ) > 1 ) {
        pop( @name_array );
        join( '/', @name_array );
    } else {
        $name;
    }
}

sub load {
    local( $size, $name ) = @_;
    $size{ $name } = $size;
    $parent = &dname( $name );
    if ( $parent ne $name ) {
        $parent{ $name } = $parent;
        @children = split( "\n", $children{ $parent } );
        push( @children, $name );
        $children{ $parent } = join( "\n", @children );
    }
}

sub do_output {
    local( $name ) = @_;

    &print_long( $size{$name}, $name );

    @children = split( "\n", $children{ $name } );

    $dot_size = $size{$name};

    $subtree_size = $dot_size;

    grep( $dot_size -= $size{$_}, @children );

    $dot_name = $name . "/.";

    push( @children, $dot_name );

    $size{$dot_name} = $dot_size;

    if ( $threshold_is_percent ) {
        @children =
            grep( ($size{$_}*100)/$total_size >= $threshold, @children );
    } else {
        @children = grep( $size{$_} >= $threshold, @children );
    }

    @children = reverse sort { $size{$a} <=> $size{$b}; } @children;

    @push_list = reverse grep( $_ ne $dot_name, @children );
    push( @subtree_list, @push_list );

    $num_components = scalar( split( '/', $name ) );

    # do short output here

    $percent_char = '%';

    foreach (@children) {
        $size = $size{$_};
        @name_array = split( '/', $_ );
        splice( @name_array, $[, $num_components );
        $_ = @name_array[$[];
        &print_short( $size, $_ );
        $percent_char = ' ';
    }
}

sub print_short {
    local( $size, $name ) = @_;
    $sub_percent = int( ($size*100)/$subtree_size );
    printf "    %3d%s %3d%s %6d    %s\n",
        int( ($size*100)/$total_size ), $percent_char,
        $sub_percent, $percent_char,
        $size,
        $name;
}

sub print_long {
    local( $size, $name ) = @_;
    print "\n";
    printf "%3d%%  %6d  %s\n",
        int( ($size*100)/$total_size ),
        $size,
        $name;
}

sub print_headers {
    print "Tot%  Blocks  Full-pathname\n";
    print "    Tot% Sub% Blocks    Child\n";
}

###############################################################

    # These next few lines are legal in both Perl and nroff.

.00;                       # finish .ig
 
'di           \" finish diversion--previous line must be blank
.nr nl 0-1    \" fake up transition to first page again
.nr % 0         \" start at page 1
'; __END__ ##### From here on it's a standard manual page #####

.TH DURF 1 "February 10, 1993"
.AT 3
.SH NAME
durf \- disk usage report formatter
.SH SYNOPSIS
.B durf [\-H] [\-t threshold[%]] [{du-args|\-}]
.SH DESCRIPTION
.I Durf
reformats the output of
.IR du\^ (1)
to make it a little easier to get an understanding of the distribution of space.
.P
The output is grouped by subtrees.
Each subtree is shown as a line like this:
.sp
.nf
\fITotal\fP%\fI   Blocks   Full-pathname\fP
.fi
.sp
which shows the position in the hierarchy and reports size
information for the subtree, followed by a group of indented lines like this:
.sp
.nf
      \fITotal\fP%\fI  Subtree\fP%\fI   Blocks   Child\fP
.fi
.sp
which reports size information for that portion of the subtree.
.I Child
is an immediate child of
.IR Full-pathname .
.IR Total %
is the percentage contribution to the total size.
.IR Subtree %
is the percentage contribution to the subtree.
.I Blocks
are disk blocks as reported by
.IR du .
The list of children is in decreasing order of size.
After the list of all the children, the next non-blank line begins the
subtree for the first child, and so on.
.P
A special fake child,
.BR . ,
is listed in each subtree (if it meets the threshold requirement).
This is simply the contribution of the directory independent of
the children.  (It is the size of the subtree less the size of the
children.)
.P
If the argument to
.I durf
is
.BR \- ,
then
.I durf
reads the standard input for the
.I du
list.
Otherwise, it runs
.IR du ,
passing it
.IR du-args .
You may pass options to
.I du
by terminating the
.I durf
options with
.BR \-\- .
.SS Options
.TP 12
\-H
Writes a usage message and dies.
.TP
\-t \fIthreshold\fP[%]
Sets a threshold for which entries
.I durf
will show.
If the threshold ends with '%', then
.I durf
will not show entries that contribute less than
.I threshold
percent to the total space.
Otherwise, the threshold is in terms of disk blocks.
The default is \fB\-t 1%\fP.
To get a complete listing, use \fB\-t 0\fP.
.SH AUTHOR
Jim Winters
.SH "SEE ALSO"
du(1).
.SH CAUTION
Believes
.IR du .

File: examples/durf.ml
#!/usr/local/adm/src/caml/bin/cash -r
-I /usr/local/adm/src/cash
!#
open Cash;

(* We see parent after children, so we'll backpatch the parent field. *)
type dir =
  { name : string; subtree_size : int; children : list dir; parent : mutable option dir }
;

value dname n =
  let slash_i = Env_3_11.internal_rindex '/' n (String.length n) in
  if slash_i < 1 then n else String.sub n 0 slash_i
;

value get_children name last =
  loop [last] where rec loop children =
    fun
    [ [last :: roots] when name = dname last.name -> loop [last :: children] roots
    | roots -> (roots, children) ]
;


value load =
  let rex = Pcre.regexp "^\s*(\d+)\s+(\S.*)$"
  and novalues = (-1, "") in
  fun line roots ->
    match
      try
        let substrs = Pcre.extract ~rex ~full_match:False line in
        (int_of_string substrs.(0), substrs.(1))
      with
      [ Not_found ->
          do { prerr_string "Can't understand line:\n"; prerr_endline line; novalues } ]
    with
    [ (-1, _) -> roots
    | (size, name) ->
        match roots with
        [ [last :: roots] when name = dname last.name ->
            let (newroots, children) = get_children name last roots in
            let thisdir =
              {name = name; subtree_size = size; children = children; parent = None}
            in
            do {
              (* Backpatch children. *)
              (let sdir = Some thisdir in
               List.iter (fun child -> child.parent := sdir) children);
              [thisdir :: newroots]
            }
        | _ ->
            (* a new root or a last's brother. *)
            [{name = name; subtree_size = size; children = []; parent = None} ::
             roots] ] ]
;

value print_long total_size size name =
  Printf.printf "\n%3d%%  %6d  %s\n" (100 * size / total_size) size name
;

value print_short total_size subtree_size percent_char size name =
  let sub_percent = size * 100 / subtree_size
  and percent = size * 100 / total_size in
  Printf.printf "    %3d%c %3d%c %6d    %s\n" percent percent_char sub_percent
    percent_char size name
;

value rec do_subtree apply_threshold total_size dir =
  let subtree_size = dir.subtree_size in
  do {
    print_long total_size subtree_size dir.name;
    let dot_size =
      List.fold_left (fun sz child -> sz - child.subtree_size) subtree_size dir.children
    in
    let dot_name = dir.name ^ "/." in
    let dot = {name = dot_name; subtree_size = dot_size; children = []; parent = None} in
    let children =
      List.sort (fun a b -> compare b.subtree_size a.subtree_size)
        (List.filter apply_threshold [dot :: dir.children])
    in
    (* XXX. *)
    let num_components = List.length (split_file_name dir.name) in
    ignore
      (let print_short = print_short total_size subtree_size in
       List.fold_left
         (fun percent_char child ->
            let name = List.nth (split_file_name child.name) num_components in
            do { print_short percent_char child.subtree_size name; ' ' })
         '%' children);
    List.iter (do_subtree apply_threshold total_size) (List.filter (\<> dot) children)
  }
;

value disk_usage_report_formatter threshold threshold_is_percent du_chan =
  (* # Load input, while establishing list of roots. *)
  let root_list = fold_in_channel du_chan read_line load [] in
  (* # Calculate total size. *)
  let total_size = List.fold_left (fun tot root -> tot + root.subtree_size) 0 root_list in
  do {
    (* print headers. *)
    print_string "Tot%  Blocks  Full-pathname\n";
    print_string "    Tot% Sub% Blocks    Child\n";
    let threshold =
      if threshold_is_percent then threshold * total_size / 100 else threshold
    in
    let apply_threshold child = child.subtree_size >= threshold in
    (* Output formatted tree for each root. *)
    List.iter (do_subtree apply_threshold total_size) root_list
  }
;

(* # Prepare file handle. *)
value mkdu use_stdin du_args =
  if use_stdin then do {
    set_chan_buffering_in stdin Block; (* probably a file, anyway. *)
    stdin
  }
  else run_with_in_channel (fun () -> exec_path "du" du_args)
;

value disk_usage_report_formatter_from_sh () =
  let threshold = ref 1
  and threshold_is_percent = ref True
  and use_stdin = ref False
  and du_args = ref [] in
  let spec =
    [("-t",
      Arg.String
        (fun s ->
           let lci = pred (String.length s) in
           do {
             threshold_is_percent.val := s.[lci] = '%';
             threshold.val :=
               int_of_string (if threshold_is_percent.val then String.sub s 0 lci else s)
           }),
      "threshold[%]\n\t Only show directories at or above threshold blocks.\n\t \
       With \"%\", only show directories at or above threshold percent\n\t \
       of the total.  Default is \"1%\"");
     ("-", Arg.Set use_stdin, "\t Read standard input for 'du' listing.")]
  and usage =
    "Usage: durf [-t #[%]] [{du-args|-}]\nLike 'du', but formatted differently."
  in
  do {
    Arg.parse spec (fun arg -> du_args.val := [arg :: du_args.val]) usage;
    disk_usage_report_formatter threshold.val threshold_is_percent.val
      (mkdu use_stdin.val du_args.val)
  }
;

if Sys.interactive.val then() else disk_usage_report_formatter_from_sh ();
